src/throttle/amoc_throttle.erl

```diff
@@ -3,57 +3,63 @@
 -module(amoc_throttle).
 
 %% API
--export([start/2,
-         start/3,
-         start/4,
-         send/3,
-         send/2,
-         send_and_wait/2,
-         wait/1,
-         run/2,
-         pause/1,
-         resume/1,
-         change_rate/3,
-         change_rate_gradually/6,
-         stop/1]).
-
--deprecated([{send_and_wait, 2, "use wait/1 instead"}]).
-
--define(DEFAULT_NO_PROCESSES, 10).
--define(DEFAULT_INTERVAL, 60000). %% one minute
--define(NONNEG_INT(N), (is_integer(N) andalso N >= 0)).
--define(POS_INT(N), (is_integer(N) andalso N > 0)).
+-export([start/2, stop/1,
+         send/2, send/3, wait/1,
+         run/2, pause/1, resume/1, unlock/1,
+         change_rate/2, change_rate_gradually/2]).
 
 -type name() :: atom().
--type rate() :: pos_integer().
--type interval() :: non_neg_integer().
-%% In milliseconds, defaults to 60000 (one minute) when not given.
-%% An interval of 0 means no delay at all, only the number of simultaneous executions will be
-%% controlled, which corresponds to the number of processes started
--export_type([name/0, rate/0, interval/0]).
-
-%% @see start/4
--spec start(name(), rate()) -> ok | {error, any()}.
-start(Name, Rate) ->
-    start(Name, Rate, ?DEFAULT_INTERVAL).
+%% Atom representing the name of the throttle.
+
+-type rate() :: infinity | non_neg_integer().
+%% Number of events per given `t:interval/0', or
+%% - infinity for effectively unlocking all throttling,
+%% - zero for effectively pausing the throttle.
+
+-type interarrival() :: infinity | non_neg_integer().
+%% Time in milliseconds between two events, or
+%% - infinity for effectively pausing the throttle,
+%% - zero for effectively unlocking all throttling.
+
+-type interval() :: pos_integer().
+%% In milliseconds, defaults to 60000 (one minute).
```

Comment:
type should be non-negative integer, with zero for limiting parallel execution

Reply:
Huh, I changed these things a long time ago, let me refresh my thoughts...

The thing is that I don't think this is a good way to "limit parallel execution". What we were really doing was only limiting parallel "starting" of execution, but then there's the question of what if the executing action is long-timed? This wasn't waiting for any previous one to finish before starting a new one. See for example the difference between [`timer:apply_interval/4`](https://www.erlang.org/doc/apps/stdlib/timer.html#apply_interval/4) and [`timer:apply_repeatedly/4`](https://www.erlang.org/doc/apps/stdlib/timer.html#apply_repeatedly/4).

It also has the problem that it is an implementation detail we're leaking here. If you remember, we had this issue with not being able to set the rate too low or too high, because then it conflicts with the number of workers and makes the throttling wrong. When we gradually change a config, we can only change so much before the "parallelism" breaks the throttling calculations.

So here I'm taking a very complex change to actually hide entirely the number of workers and be able to enable or disable any of them as needed. And it is all tested.

If what we actually want is to limit parallel execution, then we'd need to consider whether waiting for actions to end or not as the `timer` module does, and that's an entirely different can of worms here.

Reply:
we were not starting any new action before any existing action finished. I'd like to keep this semantics.

Reply:
Same thing, what if, as the `timer` module does, I wanted to keep track of starting points and not full execution? And again, we were doing one or the other based on whether the given `fun` actually does stuff (for which the throttle blocks) or sends a message to allow someone else to do stuff (for which the throttle doesn't know).

This design is leaky, tremendously leaky. And the code where we could not have too high or too low rates because of the number of processes is not just more leaky but also ugly and limiting, it wouldn't allow me to run gradual changes to the edges of the config, which is something I really have a problem with.

And giving an interval of zero an ad-hoc meaning of "parallel executions" is even uglier, now we're mixing _three_ different ways to configure stuff (rate, interarrival, and parallelisation) with _two_ configuration keys. If throttle is a rate/interval computation, the intuitive thing to happen when interval gets lower and lower is for the throttle to get bigger and bigger, the fact that gets "capped" at a (paralellism) limit is an implementation detail we're not supposed to leak.

If what we want is a parallelism mechanism, I would actually give it a properly different API (more keys to be excluding with rate or interarrival as these are already excluding each other).

Reply:
Hmm... overloading a different API is all too ugly, but I'm going to experiment with adding a proper API for this. What I want the most is to get rid of the code that limits what rates can I set based on the parallelism given.